kvserver: add basic FlushLockTable request

stevendanna opened the pull request:
This adds a new request FlushLockTableRequest. When issues, any
unreplicated locks in the request header's span will be written as
replicated locks to disk and removed from the in-memory lock table.s

The goal of this request is two fold:

1) By modelling this as a request, we can easily issue flush
   operations during KVNemesis at random. This will help us uncover
   problems with our current flush semantics.

2) We may use this flush request from a new queue that will
  periodically ensure that the size of the lock table is
  under control.

Epic: [CRDB-42764](https://cockroachlabs.atlassian.net/browse/CRDB-42764)

Release note: None

stevendanna committed:
kvserver: add basic FlushLockTable request

This adds a new request FlushLockTableRequest. When issued, any
unreplicated locks in the request header's span will be written as
replicated locks to disk and removed from the in-memory lock table.

This is an unsplittable request that can't span range boundaries.

The goal of this request is two fold:

1) By modelling this as a request, we can easily issue flush
   operations during KVNemesis at random. This will help us uncover
   problems with our current flush semantics.

2) We may use this flush request from a new queue that will
   periodically ensure that the size of the lock table is
   under control.

Epic: CRDB-42764

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141821)
<!-- Reviewable:end -->

stevendanna committed:
kvnemesis: add FlushLockTableOperation

This adds FlushLockTable to the set of KVNemsis operations. It doesn't
add any particularly interesting validations.

Epic: none
Release note: None

stevendanna commented:
@arulajmani Gentle ping here

arulajmani reviewed (commented):
Mostly looks good. Sorry for not getting to this last week.

> Reviewed 25 of 25 files at r2, 8 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-:-OMlq611EtRu6xE4RVro:b-dq211m)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-OMldnzODBFKG2GI2_Dr:-OMldnzP8J510t17U90s:b-p0cp1j):*
nit: "when issued"
___
*[`-- commits` line 5 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-OMleY7fEf_JybfY5osQ:-OMleY7fEf_JybfY5osR:b-bse3l1):*
Should we say something about how this span can't span range boundaries, given we're setting the `isUnsplitable` flag below? 
___
*[`-- commits` line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-OMldsI_-wdzaussqG3c:-OMldsI_-wdzaussqG3d:b-35901):*
nit: s/table.s/tables/g
___
*[`-- commits` line 26 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-OMlpS3F1wFT2GDjyd2_:-OMlpS3F1wFT2GDjyd2a:bfew7hs):*
Would it make sense to do some validation here? One way to do so would be to chain a `QueryIntentRequest` at some point before committing the transaction to validate the replicated lock, with some probability[1].

We don't have to do this in this patch, if we don't want to. But maybe let's file an issue to track this, if you're on board?

----

[1] I don't think we want to always do this, as we still want to test the case where the pipeliner has to chain a `QueryIntentRequest`.
___
*[`pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-OMlo-riEwso2rHshJBq:-OMlo-riEwso2rHshJBr:b8dv193) ([raw file](https://github.com/cockroachdb/cockroach/blob/169a312e9077dda5685e55960249711de2041bf9/pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go#L21)):*
> ```Go
> 	// TODO(review): We want DefaultDeclareKeys here because we don't want to have
> 	// to wait out a long running transaction. But, we need a way to verify that
> 	// the latches are enough to provide the isolation we need here. I'm hoping
> ```

I think this gives us the isolation we need against `ResolveIntentRequests` that are running concurrently, as we're declaring a write latch over the entire span with `DefaultDeclareKeys`. 

Using `DefaultDeclareKeys` here means that non-locking readers would be able to run concurrently here. I think that's okay, if we fix the locking issue below. Still, want to give this a bit more thought.
___
*[`pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go` line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-OMlm_hX4rtaLsT47_Cj:-OMlm_hX4rtaLsT47_Ck:b-x2qr8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/169a312e9077dda5685e55960249711de2041bf9/pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go#L36)):*
> ```Go
> 	// TODO(ssd): Allow the caller to limit how many locks we write out.
> 	locksToFlush := make([]roachpb.LockAcquisition, 0)
> 	cArgs.EvalCtx.GetConcurrencyManager().ExportUnreplicatedLocks(args.Span(), func(l *roachpb.LockAcquisition) {
> ```

Unrelated to your PR, but I noticed we're not locking `keyLocks.mu` in `ExporeUnreplicatedLocks` -- is that intentional?

I think it works for now, because both the places we use it in right now -- `OnRangeSubsumeEval` and `OnRangeLeaseTransferEval` declare latches over `declareAllKeys`. If we're not doing that here, I think there could be a race, right?


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
> Reviewed 2 of 25 files at r2, 20 of 20 files at r4, 8 of 8 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141821)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @stevendanna, and @sumeerbhola)

___
*[`pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-OMlo-riEwso2rHshJBq:-OO96BhV0hjeNjsFK--a:b-qb1hfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/169a312e9077dda5685e55960249711de2041bf9/pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go#L21)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I think this gives us the isolation we need against `ResolveIntentRequests` that are running concurrently, as we're declaring a write latch over the entire span with `DefaultDeclareKeys`. 

Using `DefaultDeclareKeys` here means that non-locking readers would be able to run concurrently here. I think that's okay, if we fix the locking issue below. Still, want to give this a bit more thought.
</blockquote></details>

`DefaultDeclareKeys` declares keys in the mvcc space, right? Isn't the lock table where we write here in the local non-mvcc space?
___
*[`pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go` line 40 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-OO9AapAEDtmqNVt9Y8h:-OO9AapAEDtmqNVt9Y8i:bg583to) ([raw file](https://github.com/cockroachdb/cockroach/blob/28a307025bedd7b08c7c544085acc5f63b5cddf2/pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go#L40)):*
> ```Go
> 		if err := storage.MVCCAcquireLock(ctx, rw,
> 			&l.Txn, l.IgnoredSeqNums, l.Strength, l.Key,
> 			cArgs.Stats, 0, 0); err != nil {
> ```

What happens if a durable lock was already acquired? This should return a `LockConflictError`, but that's ok, right? Do we want to return, or just keep going with the remaining locks?
___
*[`pkg/kv/kvnemesis/validator.go` line 739 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-OO9DZ1w4-ON3htxUqwG:-OO9DZ1w4-ON3htxUqwH:b-axnndv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c707bec58ccbf97de7f4fea429b08cc08e2db24/pkg/kv/kvnemesis/validator.go#L739)):*
> ```Go
> 	// concurrent writes are applied by the time it completes. Maybe later.
> 	case *FlushLockTableOperation:
> 		// TODO(ssd): Should this be true? Right now it needs
> ```

I thought the timestamp is used for serializability validation. It doesn't look like we want to serialize this with other operations, right?


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141821)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @miraradeva, and @sumeerbhola)

___
*[`-- commits` line 5 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-OMleY7fEf_JybfY5osQ:-OMpeXx4-x-utCt9USQ-:b-896fix):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we say something about how this span can't span range boundaries, given we're setting the `isUnsplitable` flag below?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-OMlo-riEwso2rHshJBq:-OOSSA8k5mDSOj0SRLnZ:bcnq8u3) ([raw file](https://github.com/cockroachdb/cockroach/blob/169a312e9077dda5685e55960249711de2041bf9/pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go#L21)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

`DefaultDeclareKeys` declares keys in the mvcc space, right? Isn't the lock table where we write here in the local non-mvcc space?
</blockquote></details>

Hrm, good point. At first I was going to say "yes that is OK because anyone taking a lock would need a write latch on the mvcc keyspace," but I think that doesn't account for Shared locks. 
___
*[`pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go` line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-OMlm_hX4rtaLsT47_Cj:-OMlqt9QCo5SfsthJRDw:b-xds5km) ([raw file](https://github.com/cockroachdb/cockroach/blob/169a312e9077dda5685e55960249711de2041bf9/pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go#L36)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Unrelated to your PR, but I noticed we're not locking `keyLocks.mu` in `ExporeUnreplicatedLocks` -- is that intentional?

I think it works for now, because both the places we use it in right now -- `OnRangeSubsumeEval` and `OnRangeLeaseTransferEval` declare latches over `declareAllKeys`. If we're not doing that here, I think there could be a race, right?
</blockquote></details>

Ah, good catch. 
___
*[`pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go` line 40 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-OO9AapAEDtmqNVt9Y8h:-OOSQGvQ50qmcxOBtwz_:b7hu07t) ([raw file](https://github.com/cockroachdb/cockroach/blob/28a307025bedd7b08c7c544085acc5f63b5cddf2/pkg/kv/kvserver/batcheval/cmd_flush_lock_table.go#L40)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

What happens if a durable lock was already acquired? This should return a `LockConflictError`, but that's ok, right? Do we want to return, or just keep going with the remaining locks?
</blockquote></details>

I think we should write some tests for this case. My "off the top of my head" answer is that if a transaction has this unreplicated lock then any replicated lock must either also be held by our transaction or be a share lock that is compatible with our lock. But, that is a bit hand-wavy and may be a result of vacation-brain.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 20 files at r4, 19 of 19 files at r6, 8 of 8 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141821#-:-OThfgl_2xFu5mSvy975:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @miraradeva, @stevendanna, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

stevendanna commented:
bors r=arulajmani
